File: app/src/main/java/com/lifedawn/bestweather/weathers/view/DetailDoubleTemperatureView.java
package com.lifedawn.bestweather.weathers.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.PointF;
import android.text.TextPaint;
import android.util.TypedValue;
import android.view.View;

import com.lifedawn.bestweather.R;
import com.lifedawn.bestweather.theme.AppTheme;
import com.lifedawn.bestweather.weathers.FragmentType;

import java.util.ArrayList;
import java.util.List;

public class DetailDoubleTemperatureView extends View implements ICleaner {
	private final FragmentType fragmentType;

	private final int viewWidth;
	private final int viewHeight;
	private final int columnWidth;
	private final int maxTemp;
	private final int minTemp;
	private final TextPaint tempPaint;
	private final Paint linePaint;
	private final Paint maxCirclePaint;
	private final Paint minCirclePaint;
	private final int circleRadius;
	private final String tempUnit;

	private List<Integer> maxTempList = new ArrayList<>();
	private List<Integer> minTempList = new ArrayList<>();

	public DetailDoubleTemperatureView(Context context, FragmentType fragmentType, int viewWidth, int viewHeight, int columnWidth, List<Integer> minTempList,
	                                   List<Integer> maxTempList) {
		super(context);
		this.fragmentType = fragmentType;

		this.viewWidth = viewWidth;
		this.viewHeight = viewHeight;
		this.columnWidth = columnWidth;

		circleRadius = (int) getResources().getDimension(R.dimen.circleRadiusInDoubleTemperature);

		tempPaint = new TextPaint();
		tempPaint.setTextAlign(Paint.Align.CENTER);
		tempPaint.setTextSize(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, 14f, getResources().getDisplayMetrics()));
		tempPaint.setColor(AppTheme.getTextColor(fragmentType));

		linePaint = new Paint();
		linePaint.setAntiAlias(true);
		linePaint.setStyle(Paint.Style.STROKE);
		linePaint.setStrokeWidth(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 1.3f, getResources().getDisplayMetrics()));
		linePaint.setColor(AppTheme.getTextColor(fragmentType));

		maxCirclePaint = new Paint();
		maxCirclePaint.setAntiAlias(true);
		maxCirclePaint.setStyle(Paint.Style.FILL);
		maxCirclePaint.setColor(Color.RED);

		minCirclePaint = new Paint();
		minCirclePaint.setAntiAlias(true);
		minCirclePaint.setStyle(Paint.Style.FILL);
		minCirclePaint.setColor(Color.BLUE);

		this.minTempList.addAll(minTempList);
		this.maxTempList.addAll(maxTempList);

		int max = Integer.MIN_VALUE;
		int min = Integer.MAX_VALUE;
		int maxTemp = 0;
		int minTemp = 0;

		for (int i = 0; i < this.minTempList.size(); i++) {
			maxTemp = this.maxTempList.get(i);
			minTemp = this.minTempList.get(i);

			// 최대,최소 기온 구하기
			if (maxTemp >= max) {
				max = maxTemp;
			}
			if (minTemp <= min) {
				min = minTemp;
			}
		}
		this.maxTemp = max;
		this.minTemp = min;

		tempUnit = context.getString(R.string.degree_symbol);

		setWillNotDraw(false);
	}

	public void setTempTextSize(int textSizeSp) {
		tempPaint.setTextSize(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, textSizeSp, getResources().getDisplayMetrics()));
	}

	public void setTextColor(int textColor) {
		tempPaint.setColor(textColor);
	}

	@Override
	protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
		setMeasuredDimension(viewWidth, viewHeight);
	}

	@Override
	protected void onLayout(boolean changed, int l, int t, int r, int b) {
		super.onLayout(changed, l, t, r, b);
	}

	@Override
	protected void onDraw(Canvas canvas) {
		drawGraph(canvas);
	}

	private void drawGraph(Canvas canvas) {
		// 텍스트의 높이+원의 반지름 만큼 뷰의 상/하단에 여백을 설정한다.
		final float TEXT_HEIGHT = tempPaint.descent() - tempPaint.ascent();
		final float TEXT_ASCENT = -tempPaint.ascent();

		final float VIEW_HEIGHT = getHeight() - ((TEXT_HEIGHT + circleRadius) * 2);
		final float SPACING = ((VIEW_HEIGHT) / (maxTemp - minTemp)) / 10f;

		int min = 0;
		int max = 0;
		float x = 0f;
		float minY = 0f;
		float maxY = 0f;

		PointF lastMinColumnPoint = new PointF();
		PointF lastMaxColumnPoint = new PointF();

		float[] minCircleXArr = new float[minTempList.size()];
		float[] minCircleYArr = new float[minTempList.size()];
		float[] maxCircleXArr = new float[minTempList.size()];
		float[] maxCircleYArr = new float[minTempList.size()];

		List<PointF> minLinePointList = new ArrayList<>();
		List<PointF> maxLinePointList = new ArrayList<>();

		final int tempsCount = minTempList.size();


		for (int index = 0; index < tempsCount; index++) {
			min = minTempList.get(index);
			max = maxTempList.get(index);

			x = columnWidth / 2f + columnWidth * index;
			minY = (10f * (maxTemp - min)) * SPACING + TEXT_HEIGHT + circleRadius;
			maxY = (10f * (maxTemp - max)) * SPACING + TEXT_HEIGHT + circleRadius;

			canvas.drawText(minTempList.get(index).toString() + tempUnit, x, minY + circleRadius + TEXT_HEIGHT, tempPaint);
			canvas.drawText(maxTempList.get(index).toString() + tempUnit, x, maxY - circleRadius - TEXT_HEIGHT + TEXT_ASCENT, tempPaint);

			lastMinColumnPoint.set(x, minY);
			lastMaxColumnPoint.set(x, maxY);

			minCircleXArr[index] = x;
			minCircleYArr[index] = minY;
			maxCircleXArr[index] = x;
			maxCircleYArr[index] = maxY;

			minLinePointList.add(new PointF(lastMinColumnPoint.x, lastMinColumnPoint.y));
			maxLinePointList.add(new PointF(lastMaxColumnPoint.x, lastMaxColumnPoint.y));
		}

		PointF[] minPoints1 = new PointF[tempsCount];
		PointF[] minPoints2 = new PointF[tempsCount];

		PointF[] maxPoints1 = new PointF[tempsCount];
		PointF[] maxPoints2 = new PointF[tempsCount];

		Path minPath = new Path();
		Path maxPath = new Path();
		minPath.moveTo(minLinePointList.get(0).x, minLinePointList.get(0).y);
		maxPath.moveTo(maxLinePointList.get(0).x, maxLinePointList.get(0).y);

		for (int i = 1; i < tempsCount; i++) {
			minPoints1[i] = new PointF((minLinePointList.get(i).x + minLinePointList.get(i - 1).x) / 2, minLinePointList.get(i - 1).y);
			minPoints2[i] = new PointF((minLinePointList.get(i).x + minLinePointList.get(i - 1).x) / 2, minLinePointList.get(i).y);

			maxPoints1[i] = new PointF((maxLinePointList.get(i).x + maxLinePointList.get(i - 1).x) / 2, maxLinePointList.get(i - 1).y);
			maxPoints2[i] = new PointF((maxLinePointList.get(i).x + maxLinePointList.get(i - 1).x) / 2, maxLinePointList.get(i).y);

			minPath.cubicTo(minPoints1[i].x, minPoints1[i].y, minPoints2[i].x, minPoints2[i].y, minLinePointList.get(i).x,
					minLinePointList.get(i).y);
			maxPath.cubicTo(maxPoints1[i].x, maxPoints1[i].y, maxPoints2[i].x, maxPoints2[i].y, maxLinePointList.get(i).x,
					maxLinePointList.get(i).y);
		}

		canvas.drawPath(minPath, linePaint);
		canvas.drawPath(maxPath, linePaint);

		for (int i = 0; i < minCircleXArr.length; i++) {
			canvas.drawCircle(minCircleXArr[i], minCircleYArr[i], circleRadius, minCirclePaint);
			canvas.drawCircle(minCircleXArr[i], maxCircleYArr[i], circleRadius, maxCirclePaint);
		}

	}


	@Override
	public void clear() {
		maxTempList.clear();
		minTempList.clear();
	}
}

File: app/src/main/java/com/lifedawn/bestweather/ui/weathers/view/DetailDoubleTemperatureView.kt
package com.lifedawn.bestweather.ui.weathers.view

import android.content.Context
import android.graphics.*
import android.text.TextPaint
import android.util.TypedValue
import android.view.View
import com.lifedawn.bestweather.R
import com.lifedawn.bestweather.ui.theme.AppTheme.getTextColor
import com.lifedawn.bestweather.ui.weathers.FragmentType

class DetailDoubleTemperatureView(
    context: Context,
    private val fragmentType: FragmentType,
    private val viewWidth: Int,
    private val viewHeight: Int,
    private val columnWidth: Int,
    minTempList: List<Int>?,
    maxTempList: List<Int>?
) : View(context), ICleaner {
    private val maxTemp: Int
    private val minTemp: Int
    private val tempPaint: TextPaint
    private val linePaint: Paint
    private val maxCirclePaint: Paint
    private val minCirclePaint: Paint
    private val circleRadius: Int
    private val tempUnit: String
    private val maxTempList: MutableList<Int> = ArrayList()
    private val minTempList: MutableList<Int> = ArrayList()

    init {
        circleRadius = resources.getDimension(R.dimen.circleRadiusInDoubleTemperature).toInt()
        tempPaint = TextPaint()
        tempPaint.textAlign = Paint.Align.CENTER
        tempPaint.textSize = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, 14f, resources.displayMetrics)
        tempPaint.color = getTextColor(fragmentType)
        linePaint = Paint()
        linePaint.isAntiAlias = true
        linePaint.style = Paint.Style.STROKE
        linePaint.strokeWidth = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 1.3f, resources.displayMetrics)
        linePaint.color = getTextColor(fragmentType)
        maxCirclePaint = Paint()
        maxCirclePaint.isAntiAlias = true
        maxCirclePaint.style = Paint.Style.FILL
        maxCirclePaint.color = Color.RED
        minCirclePaint = Paint()
        minCirclePaint.isAntiAlias = true
        minCirclePaint.style = Paint.Style.FILL
        minCirclePaint.color = Color.BLUE
        this.minTempList.addAll(minTempList!!)
        this.maxTempList.addAll(maxTempList!!)
        var max = Int.MIN_VALUE
        var min = Int.MAX_VALUE
        var maxTemp = 0
        var minTemp = 0
        for (i in this.minTempList.indices) {
            maxTemp = this.maxTempList[i]
            minTemp = this.minTempList[i]

            // 최대,최소 기온 구하기
            if (maxTemp >= max) {
                max = maxTemp
            }
            if (minTemp <= min) {
                min = minTemp
            }
        }
        this.maxTemp = max
        this.minTemp = min
        tempUnit = context.getString(R.string.degree_symbol)
        setWillNotDraw(false)
    }

    fun setTempTextSize(textSizeSp: Int) {
        tempPaint.textSize =
            TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, textSizeSp.toFloat(), resources.displayMetrics)
    }

    fun setTextColor(textColor: Int) {
        tempPaint.color = textColor
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        setMeasuredDimension(viewWidth, viewHeight)
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        super.onLayout(changed, l, t, r, b)
    }

    override fun onDraw(canvas: Canvas) {
        drawGraph(canvas)
    }

    private fun drawGraph(canvas: Canvas) {
        // 텍스트의 높이+원의 반지름 만큼 뷰의 상/하단에 여백을 설정한다.
        val TEXT_HEIGHT = tempPaint.descent() - tempPaint.ascent()
        val TEXT_ASCENT = -tempPaint.ascent()
        val VIEW_HEIGHT = height - (TEXT_HEIGHT + circleRadius) * 2
        val SPACING = VIEW_HEIGHT / (maxTemp - minTemp) / 10f
        var min = 0
        var max = 0
        var x = 0f
        var minY = 0f
        var maxY = 0f
        val lastMinColumnPoint = PointF()
        val lastMaxColumnPoint = PointF()
        val minCircleXArr = FloatArray(minTempList.size)
        val minCircleYArr = FloatArray(minTempList.size)
        val maxCircleXArr = FloatArray(minTempList.size)
        val maxCircleYArr = FloatArray(minTempList.size)
        val minLinePointList: MutableList<PointF> = ArrayList()
        val maxLinePointList: MutableList<PointF> = ArrayList()
        val tempsCount = minTempList.size
        for (index in 0 until tempsCount) {
            min = minTempList[index]
            max = maxTempList[index]
            x = columnWidth / 2f + columnWidth * index
            minY = 10f * (maxTemp - min) * SPACING + TEXT_HEIGHT + circleRadius
            maxY = 10f * (maxTemp - max) * SPACING + TEXT_HEIGHT + circleRadius
            canvas.drawText(minTempList[index].toString() + tempUnit, x, minY + circleRadius + TEXT_HEIGHT, tempPaint)
            canvas.drawText(maxTempList[index].toString() + tempUnit, x, maxY - circleRadius - TEXT_HEIGHT + TEXT_ASCENT, tempPaint)
            lastMinColumnPoint[x] = minY
            lastMaxColumnPoint[x] = maxY
            minCircleXArr[index] = x
            minCircleYArr[index] = minY
            maxCircleXArr[index] = x
            maxCircleYArr[index] = maxY
            minLinePointList.add(PointF(lastMinColumnPoint.x, lastMinColumnPoint.y))
            maxLinePointList.add(PointF(lastMaxColumnPoint.x, lastMaxColumnPoint.y))
        }
        val minPoints1 = arrayOfNulls<PointF>(tempsCount)
        val minPoints2 = arrayOfNulls<PointF>(tempsCount)
        val maxPoints1 = arrayOfNulls<PointF>(tempsCount)
        val maxPoints2 = arrayOfNulls<PointF>(tempsCount)
        val minPath = Path()
        val maxPath = Path()
        minPath.moveTo(minLinePointList[0].x, minLinePointList[0].y)
        maxPath.moveTo(maxLinePointList[0].x, maxLinePointList[0].y)
        for (i in 1 until tempsCount) {
            minPoints1[i] = PointF((minLinePointList[i].x + minLinePointList[i - 1].x) / 2, minLinePointList[i - 1].y)
            minPoints2[i] = PointF((minLinePointList[i].x + minLinePointList[i - 1].x) / 2, minLinePointList[i].y)
            maxPoints1[i] = PointF((maxLinePointList[i].x + maxLinePointList[i - 1].x) / 2, maxLinePointList[i - 1].y)
            maxPoints2[i] = PointF((maxLinePointList[i].x + maxLinePointList[i - 1].x) / 2, maxLinePointList[i].y)
            minPath.cubicTo(
                minPoints1[i]!!.x, minPoints1[i]!!.y, minPoints2[i]!!.x, minPoints2[i]!!.y, minLinePointList[i].x,
                minLinePointList[i].y
            )
            maxPath.cubicTo(
                maxPoints1[i]!!.x, maxPoints1[i]!!.y, maxPoints2[i]!!.x, maxPoints2[i]!!.y, maxLinePointList[i].x,
                maxLinePointList[i].y
            )
        }
        canvas.drawPath(minPath, linePaint)
        canvas.drawPath(maxPath, linePaint)
        for (i in minCircleXArr.indices) {
            canvas.drawCircle(minCircleXArr[i], minCircleYArr[i], circleRadius.toFloat(), minCirclePaint)
            canvas.drawCircle(minCircleXArr[i], maxCircleYArr[i], circleRadius.toFloat(), maxCirclePaint)
        }
    }

    override fun clear() {
        maxTempList.clear()
        minTempList.clear()
    }
}